kvbc/src/kvbc_app_filter/kvbc_app_filter.cpp

```diff
@@ -271,48 +272,273 @@ void KvbAppFilter::readBlockRange(BlockId block_id_start,
+      if (!has_pub_offset_updated && std::find(public_event_group_ids.begin(), public_event_group_ids.end(), *r_it) !=
+                                         public_event_group_ids.end()) {
+        auto pub_it = std::find(public_event_group_ids.begin(), public_event_group_ids.end(), *r_it);
+        eg_state->public_offset = ++pub_it - public_event_group_ids.begin();
+        has_pub_offset_updated = true;
+        LOG_DEBUG(logger_,
+                  "Updated public_offset: " << eg_state->public_offset
+                                            << " public_event_group_ids size: " << public_event_group_ids.size());
+      }
+      if (!has_pvt_offset_updated && std::find(private_event_group_ids.begin(), private_event_group_ids.end(), *r_it) !=
+                                         private_event_group_ids.end()) {
+        auto pvt_it = std::find(private_event_group_ids.begin(), private_event_group_ids.end(), *r_it);
+        eg_state->private_offset = ++pvt_it - private_event_group_ids.begin();
+        has_pvt_offset_updated = true;
+        LOG_DEBUG(logger_,
+                  "Updated private_offset: " << eg_state->private_offset
+                                             << " private_event_group_ids size: " << private_event_group_ids.size());
+      }
+    }
+  }
+  // let's not return an invalid event group id
+  if (eg_state->event_group_id_batch.empty()) return std::nullopt;
+
+  // Reset the iterator so that we read from event_group_id_batch from the beginning
+  eg_state->it = eg_state->event_group_id_batch.begin();
+
+  ConcordAssertLE(eg_state->event_group_id_batch.size(), kBatchSize);
+  LOG_DEBUG(logger_, "Updated event group id list batch size: " << eg_state->event_group_id_batch.size());
+  return *eg_state->it++;
+}
+
+void KvbAppFilter::readEventGroupRange(EventGroupId event_group_id_start,
+                                       spsc_queue<KvbFilteredEventGroupUpdate> &queue_out,
+                                       const std::atomic_bool &stop_execution) {
+  uint64_t public_start = getValueFromLatestTable(kPublicEgIdKeyOldest);
+  uint64_t private_start = getValueFromLatestTable(client_id_ + "_oldest");
+  uint64_t public_end = getValueFromLatestTable(kPublicEgIdKeyNewest);
+  uint64_t private_end = getValueFromLatestTable(client_id_ + "_newest");
+  LOG_DEBUG(logger_,
+            "Reading event group range, public_start: " << public_start << " private_start: " << private_start
+                                                        << " public_end: " << public_end
+                                                        << " private_end: " << private_end);
+  if (!public_start && !private_start) {
+    std::stringstream msg;
+    msg << "Event groups do not exist for client: " << client_id_ << " yet.";
+    LOG_ERROR(logger_, msg.str());
+    throw std::runtime_error(msg.str());
+  } else if (!public_start) {
+    ConcordAssertEQ(public_start, public_end);
```

Comment:
Let's say we have public event groups 1-7 in the db. Now, we prune all of them and therefore, they cannot be queried. Meaning, `public_event_group_oldest` should be `0` (invalid). However, `public_event_group_newest` should be `7` because we need to preserve this value in order to increment it for future public event groups.

Reply:
Updated.